robocopkaka/house-renting-app
# Add methods for adding, updating, retrieving and deleting properties

#### What does this PR do?
Add Property resource
#### Description of Task to be completed?
 - add method for adding new properties
  - add middleware for validating input for properties
  - add middleware for authenticating landlords
  - add route for adding properties
  - add tests
  - add methods for updating a property
  - add methods and tests for deleting a property
  - add methods and tests for retrieving a property
  - add methods and tests for retrieving properties

misc:
  - rename User to Landlord
  - rename Landlord to LandlordController in userController.ts
  - return all errors when validating schemas
#### How should this be manually tested?
A) To create a property, send:
```
curl -L -X POST 'http://localhost:8000/properties' \
-H 'token: actual token' \
-H 'Content-Type: application/json' \
--data-raw '{
	"name": "Test Property",
	"description": "Echi",
	"propertyType": "commercial",
	"address": "here",
	"category": "rent",
	"available": "false"
}'
```

Should return:
```.json
{
    "message": "Property added successfully",
    "property": {
        "id": 17,
        "name": "Test Property",
        "description": "Echi",
        "propertyType": "commercial",
        "address": "here",
        "features": [],
        "category": "rent",
        "available": false,
        "landlordId": 1,
        "uid": "ba0fd456-86d3-40f8-9cd6-63dea745c33b"
    }
}
```

B) To update a property, send

```
curl -L -X PATCH 'http://localhost:8000/properties/11' \
-H 'token: actualToken' \
-H 'Content-Type: application/json' \
--data-raw '{
	"name": "Edited name twice",
	"address": "edited here"
}'
```

Which should return:

```.json
{
    "message": "Property updated successfully",
    "property": {
        "id": 11,
        "name": "Edited name twice",
        "propertyType": "commercial",
        "description": "Echi",
        "address": "edited here",
        "features": [],
        "uid": "66f83de6-607f-4b98-97a7-b8baecbcafcb",
        "category": "rent",
        "available": false,
        "landlordId": 1
    }
}
```

C) Retrieve all properties
Send: 

```
curl -L -X GET 'http://localhost:8000/properties'
```

Which should return:

```.json
{
    "properties": [
        {
            "id": 2,
            "name": "Test Property",
            "propertyType": "commercial",
            "description": "Echi",
            "address": "here",
            "features": null,
            "uid": "d2e029c4-3ce8-43fa-aaa4-b0df4f274445",
            "category": "rent",
            "available": false,
            "landlordId": 1,
            "createdAt": "2020-04-12T04:52:58.593Z",
            "updatedAt": "2020-04-12T04:52:58.593Z"
        },
        {
            "id": 3,
            "name": "Test Property",
            "propertyType": "commercial",
            "description": "Echi",
            "address": "here",
            "features": null,
            "uid": "2bfcaff3-9035-4596-a30c-8672d916d12f",
            "category": "rent",
            "available": false,
            "landlordId": 1,
            "createdAt": "2020-04-12T05:27:41.558Z",
            "updatedAt": "2020-04-12T05:27:41.558Z"
        }
    ],
    "message": "Properties returned successfully"
}
```

D) Return a single property
Send

```
curl -L -X GET 'http://localhost:8000/properties/2'
```

Which should return:

```.json
{
    "property": {
        "id": 2,
        "name": "Test Property",
        "propertyType": "commercial",
        "description": "Echi",
        "address": "here",
        "features": [],
        "uid": "d2e029c4-3ce8-43fa-aaa4-b0df4f274445",
        "category": "rent",
        "available": false,
        "landlordId": 1
    },
    "message": "Property returned successfully"
}
```

E) To delete a property,
Send:

```
curl -L -X DELETE 'http://localhost:8000/properties/3' \
-H 'token: actualTokens'
```

Which should return:

```.json
{
    "message": "Property deleted successfully"
}
```

#### Any background context you want to provide?
If you experience any issues running tests, connect to your test database through your terminal and try the most accepted answer here - https://stackoverflow.com/questions/3327312/how-can-i-drop-all-the-tables-in-a-postgresql-database

#### What are the relevant pivotal tracker stories? (if applicable)
https://www.pivotaltracker.com/story/show/171429253
https://www.pivotaltracker.com/story/show/172163127
https://www.pivotaltracker.com/story/show/171429387
